### DISCUSSION TITLE
백엔드 코드 컨벤션을 맞춰봅시다

### DISCUSSION THREAD

**Discussion by Kim0914** [woowacourse-teams/2023-pium#61] [뜨거운 토론]
### 클라이언트 요청에 대한 예외 처리(Exception Handling)를 어떻게 할 것인가?
`ControllerAdvice`를 사용해서 전역으로 예외 핸들링을 할 수 있습니다.

`ControllerAdvice`를 구현하는 방법이 여러가지 방법이 있습니다.

어떤 팀 컨벤션을 가져갈지 이야기 나눠보면 좋을 것 같아요 !

---

**Comment by Choi-JJunho** [comment]
## 컨벤션

### DTO

- 빌더 사용 시,  클래스 레벨에 `@Builder` , `@AllArgsConstructor(access = AccessLevel.PRIVATE)` 사용한다.
- 하나의 DTO내에서만 필요한 클래스의 경우 내부 클래스로 작성 한다. 이 때, 내부 클래스는 static으로 작성한다.

### TEST

- fixture는 영속화 하지 않도록 한다.

### Transactional

- Service 계층에서 우선적으로 클래스 레벨에는 `read-only=true` 옵션을 사용한다.
- CUD가 필요한 메서드에는 메서드 레벨로 `@Transactional` 옵션을 사용한다.

### 예외

- 표준 예외를 사용하는 것을 기본으로 한다.
- 예외 메세지를 작성할 때는 `개발자가 생각하는 문제 + 원인` 형식으로 작성한다.
    
    (ex) `해당 반려식물이 존재하지 않습니다. id = 1L`
    
- `ResponseEntityExceptionHandler` 사용
    - 우리가 예측하지 못한 예외에 대한 1차적인 처리를 해준다.
    - 스프링 MVC와 관련된 예외는 모두 GlobalExceptionHandler에서 처리한다.
    - 제공해주는 기본 메세지를 우선적으로 사용하고, 이후에 수정이 필요하면 override한다.

### 기타

- api 명세 수정 내용 프론트에 전달할 것
- controller, service 메서드 네이밍
    - 기본으로 CRUD를 사용하고 겹치거나 필요한 경우에 추가적으로 작성 
    (ex) `read()`, `readAll()`
- PetPlant
    - nextWaterDay, daySince : integer → long
- Request, Response가 담긴 dto 패키지는 서비스 패키지다.

**Comment by Choi-JJunho** [comment]
해당 내용 Wiki에 정리했습니다.
https://github.com/woowacourse-teams/2023-pium/wiki/%5BBE%5D-%EC%BD%94%EB%93%9C-%EC%BB%A8%EB%B2%A4%EC%85%98
